File: code/PCC-javathreads/src/partial/End.java
package partial;

import choco.Choco;
import choco.kernel.model.constraints.Constraint;
import partial.sat.Solver;

/**
 * Created by IntelliJ IDEA.
 * User: dave
 * Date: 10/01/12
 * Time: 14:12
 * To change this template use File | Settings | File Templates.
 */
public abstract class End {
    private String name;
    private Primitive primitive;

    public End(String name, Primitive primitive) {
        this.name = name;
        this.primitive = primitive;
    }

    public String getName() {
        return name;
    }

    public Primitive getPrimitive() {
        return primitive;
    }

    abstract boolean isConnected();
    abstract End getOther();

    abstract Constraint getConstraint();
    
    public Constraint getNoFlowConstraint() {
        // var may still not exist yet!
        return Choco.eq(Solver.getvar(primitive.getVars(),name),0);
    }

    public boolean isReady() {
        return false;
    }
}

File: code/PCC-scalaactors/src/partial/reo/End.scala
package partial.reo

import partial.solver.{Neg, Var, Constr}


/**
 * Created by IntelliJ IDEA.
 * User: jose
 * Date: 29/02/12
 * Time: 16:43
 * To change this template use File | Settings | File Templates.
 */

abstract class End(name:String, primitive:Primitive) {

    def getName = name

    def getPrimitive = primitive

    def isConnected: Boolean

    def getOther: End

    def getConstraint: Constr

    def getNoFlowConstraint: Constr = Neg(Var(name))
  
    def isReady:Boolean = false

}
